haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,203 @@
+import logging
+from pathlib import Path
+from typing import List, Optional, Union, Dict, Any
+
+from haystack.preview import ComponentError
+
+from haystack.lazy_imports import LazyImport
+from haystack.preview import Document, component, default_from_dict, default_to_dict
+
+
+logger = logging.getLogger(__name__)
+
+
+with LazyImport(message="Run 'pip install transformers[torch,sentencepiece]==4.32.1'") as torch_and_transformers_import:
+    import torch
+    from transformers import AutoModelForSequenceClassification, AutoTokenizer
+
+
+@component
+class TopPSampler:
+    """
+    Selects documents based on the cumulative probability of the similarity scores between the
+    query and the documents using top p sampling.
+
+    Top p sampling selects a subset of the most relevant data points from a larger set of data. The technique
+    involves calculating the cumulative probability of the scores of each data point, and then
+    selecting the top p percent of data points with the highest cumulative probability.
+
+    In the context of TopPSampler, the `run()` method takes in a query and a set of documents,
+    calculates the similarity scores between the query and the documents, and then filters
+    the documents based on the cumulative probability of these scores. The TopPSampler provides a
+    way to efficiently select the most relevant documents based on their similarity to a given query.
+
+    Usage example:
+    ```
+    from haystack.preview import Document
+    from haystack.preview.components.samplers import TopPSampler
+
+    sampler = TopPSampler(top_p=0.95)
+    docs = [Document(text="Paris"), Document(text="Berlin")]
+    query = "City in Germany"
+    output = sampler.run(query=query, documents=docs)
+    docs = output["documents"]
+    assert len(docs) == 1
+    assert docs[0].text == "Berlin"
+    ```
+    """
+
+    def __init__(
+        self,
+        model_name_or_path: Optional[Union[str, Path]] = None,
+        top_p: Optional[float] = 1.0,
+        score_field: Optional[str] = "similarity_score",
+        device: Optional[str] = "cpu",
+    ):
+        """
+        Creates an instance of TopPSampler.
+
+        :param model_name_or_path: Path to a pre-trained sentence-transformers model. If not specified, no document
+        scoring will be performed, and it is assumed that documents already have scores in the metadata specified under
+        the `score_field` key.
+        :param top_p: Cumulative probability threshold for filtering the documents (usually between 0.9 and 0.99).
+        :param score_field: The name of the field that should be used to store the scores in a document's metadata.
+        :param device: torch device (for example, cuda:0, cpu, mps) to limit model inference to a specific device.
+        """
+        torch_and_transformers_import.check()
```

Comment:
We don't necessarily need these dependencies though, right? If the model is not given, the deps are not needed.

Let's only check the import is the model is given.